PowerShellTools.HostService/ServiceManagement/Debugging/PowerShellDebuggingService.cs

```diff
@@ -157,29 +187,36 @@ public void AttachToRunspace(uint pid)
                 return;
             }
 
-            // Enter into to-attach process which will swap out the current runspace
+            // enter into to-attach process which will swap out the current runspace
             using (_currentPowerShell = PowerShell.Create())
             {
+                // scenario before entering, used to determine if we are local attaching
+                DebugScenario preScenario = GetDebugScenario();
+
+                _attachRequestEvent.Reset();
                 _currentPowerShell.Runspace = _runspace;
                 _currentPowerShell.AddCommand("Enter-PSHostProcess").AddParameter("Id", pid.ToString());
                 _currentPowerShell.Invoke();
 
-                // wait for invoke to finish swapping the runspaces
-                _attachRequestEvent.WaitOne(5000);
-
-                // make sure that the semaphore didn't just time out
-                if (GetDebugScenario() != DebugScenario.LocalAttach)
+                // wait for invoke to finish swapping the runspaces if you are attaching to a local process
+                if (preScenario != DebugScenario.RemoteSession)
                 {
-                    MessageBox.Show(Resources.ProcessAttachFailErrorBody, Resources.ProcessAttachFailErrorTitle, MessageBoxButton.OK, MessageBoxImage.Warning);
-                    ServiceCommon.Log("Failed to attach to running process.");
-                    return;
+                    _attachRequestEvent.WaitOne(DebugEngineConstants.AttachRequestEventTimeout);
+
+                    // scenario after entering, used to determine if semaphore timed out, should not match our pre-scenario
+                    DebugScenario postScenario = GetDebugScenario();
+
+                    // make sure that the semaphore didn't just time out
+                    if (preScenario == postScenario)
+                    {
+                        MessageBox.Show(Resources.ProcessAttachFailErrorBody, Resources.ProcessAttachFailErrorTitle, MessageBoxButton.OK, MessageBoxImage.Warning);
+                        ServiceCommon.Log("Failed to attach to running process.");
+                        return;
+                    }
                 }
 
-                // rehook up the event handling
-                _runspace.Debugger.DebuggerStop += Debugger_DebuggerStop;
-                _runspace.Debugger.BreakpointUpdated += Debugger_BreakpointUpdated;
-                _runspace.StateChanged += _runspace_StateChanged;
-                _runspace.AvailabilityChanged += _runspace_AvailabilityChanged;
+                // rehook event handlers
+                AddEventHandlers();
 
                 // debug the runspace, for the vast majority of cases the 1st runspace is the one to attach to

```

Comment:
Not Yet. I was/am going to make that issue and a few others once this code had been checked in.